library/alloc/src/slice.rs

```diff
@@ -156,8 +156,32 @@ mod hack {
     where
         T: Clone,
     {
+        struct DropGuard<'a, T> {
+            vec: &'a mut Vec<T>,
+            num_init: usize,
+        }
+        impl<'a, T> Drop for DropGuard<'a, T> {
+            fn drop(&mut self) {
+                // SAFETY:
+                // items were marked initialized in the loop below
+                unsafe {
+                    self.vec.set_len(self.num_init);
+                }
+            }
+        }
         let mut vec = Vec::with_capacity(s.len());
-        vec.extend_from_slice(s);
+        let mut guard = DropGuard { vec: &mut vec, num_init: 0 };
+        let slots = guard.vec.spare_capacity_mut();
+        for (i, b) in s.iter().enumerate().take(slots.len()) {
```

Comment:
Is this `take(slots.len())` purely for performance to convince LLVM? Otherwise it looks not necessary...

Reply:
ah yea I found without it it does a bounds check when indexing on `slots[i]`, and although `zip` is cleaner, it also did not have nice codegen. `take` seemed to work the best.

I do agree it looks really odd, but I can leave a comment marking why it's there